Add Lab with examples and instructions (#248)
openshift/special-resource-operator: lab/README.md

> # Special Resource Operator Lab

Reviewer: Can we put this file in the `doc` directory instead of `lab`?

 

The lab assumes that there is an cluster already deployed and it contains the following exercises:

> 	- Exercise 1

Reviewer: How about using `## Table of content` with 5 links (each one with a descriptive name) to each on of the exercises?

> 	- Exercise 1
>         	- SRO will be deployed along with NFD dependency for SRO v1.
>         	- Verify SRO and NFD are running.

Reviewer: ```suggestion
[Deploy SRO with the necessary dependency NFD and verify that both are running](link to the relevant section)
```

> 	- Exercise 2
> 		- Simple kmod example used as a deployment exercise.
>  		- Verify simple kmod is deployed and running.

Reviewer: ```suggestion
[Deploy a simple-kmod](link to the relevant section)
```

> 	- Exercise 3
> 		- More advanced recipe will be used along with DTK for the following:
> 		- Use DTK to build a driver container image
> 		- Create a recipe for SRO which will deploy the newly created driver image from the local registry
> 		- Verify the Daemonset is running on the needed nodes

Reviewer: ```suggestion
[Deploy Intel's Ice driver-container](link to the relevant section)
```

We can add the removed text as part of the relevant section if needed.

> 	- Exercise 4
> 		- Most advanced recipe to include deploying a kernel module built using DTK and deploying a device plug-in
> 		- Create a recipe which deploys a kmod along with a kubernetes device plug-in
> 		- Verify both the Daemonset and device plug-in are deployed and running in cluster

Reviewer: ```suggestion
Deploy a simple driver-container as well as a device plugin
```

> 	- Exercise 5
> 		- Troubleshoot failure to load a kmod

Reviewer: ```suggestion
[Troubleshoot failures to load a driver-container](link to the relevant section)
```

## EXERCISE 1

> ** Deploy SRO with the necessary dependency NFD and verify that both are running

Reviewer: ```suggestion
** Deploy SRO with the necessary dependency NFD and verify that both are running**
```

> ## EXERCISE 1
> 
> ** Deploy SRO with the necessary dependency NFD and verify that both are running

Reviewer: ```suggestion
## Deploy SRO with the necessary dependency NFD and verify that both are running
 ```

```

After a while we should see Special Resource Operator controller running and also its dependency Node Feature Discovery :

```shell
[root@ebelarte lab]# oc get po
NAME                                                   READY   STATUS    RESTARTS   AGE
nfd-controller-manager-8c9585895-8xqmd                 2/2     Running   0          18m
special-resource-controller-manager-56b978fc6d-fhkgs   2/2     Running   0          19m
[root@ebelarte lab]#
```

 
### Using Openshift console
Login the console and go to Operators -> OperatorHub

![OperatorHub](https://github.com/enriquebelarte/sro-lab/blob/main/images/image3.png)

In the search box, type special:

![OperatorHub](https://github.com/enriquebelarte/sro-lab/blob/main/images/image2.png)

![OperatorHub](https://github.com/enriquebelarte/sro-lab/blob/main/images/image4.png)

Choose Special Resource Operator provided by Red Hat and click the Install button:

![OperatorHub](https://github.com/enriquebelarte/sro-lab/blob/main/images/image5.png)

![OperatorHub](https://github.com/enriquebelarte/sro-lab/blob/main/images/image6.png)

![OperatorHub](https://github.com/enriquebelarte/sro-lab/blob/main/images/image7.png)

> ## EXERCISE 2
> 
> 
> **Deploy simple-kmod example**

Reviewer: ```suggestion
## Deploy simple-kmod example
```

    buildArgs:
    - name: "KMODVER"
      value: "SRO"
  driverContainer:
    source:
      git:
        ref: "master"
        uri: "https://github.com/openshift-psap/simple-kmod.git"
```

Finally we should create the above SR in the cluster:

```shell
[root@ebelarte chart]# oc create -f simple-kmod-sr.yaml 
specialresource.sro.openshift.io/simple-kmod created

```

After a few minutes we could check if pods are running in simple-kmod project. As we can see below the build was done and the driver container is running:

```shell
[root@ebelarte chart]# oc get po
NAME                                                  READY   STATUS      RESTARTS   AGE
simple-kmod-driver-build-e383247e62b56585-1-build     0/1     Completed   0          9m1s
simple-kmod-driver-container-e383247e62b56585-jkh65   1/1     Running     0          9m56s
[root@ebelarte chart]# 
```

In order to make sure that the kernel modules are effectively loaded we could use the `lsmod` command:

```shell
[root@ebelarte chart]# oc exec -it simple-kmod-driver-container-e383247e62b56585-jkh65 -- lsmod | grep simple
simple_procfs_kmod     16384  0
simple_kmod            16384  0
[root@ebelarte chart]# 
```

> ## EXERCISE 3

Reviewer: ```suggestion
## [Deploy Intel's Ice driver-container](link to the relevant section)
```

**Using DTK to build a driver container image**

> Driver Toolkit a.k.a. DTK is a container image which can be used as a base image to build out-of-tree driver containers as it has all the required dependencies to do so.

Reviewer: ```suggestion
[driver-toolkit](lint to the github repo) a.k.a. DTK is a container image which can be used as a base image to build out-of-tree driver containers as it has all the required dependencies to do so.
```

```

Create the SpecialResource definition:

### `ice-kmod-sr.yaml`

```yaml
apiVersion: sro.openshift.io/v1beta1
kind: SpecialResource
metadata:
  name: ice-kmod
spec:
  namespace: ice-kmod
  chart:
    name: ice-kmod
    version: 1.8.9
    repository:
      name: ice-kmod
      url: cm://ice-kmod/ice-kmod-chart
  set:
    kind: Values
    apiVersion: sro.openshift.io/v1beta1
    kmodNames: ["ice"]
    driverContainerImage: ice-kmod-driver-container
    driverContainerVersion: 1.8.9
    buildArgs:
    - name: "KMODVER"
      value: "1.8.9"
```

Then create the SpecialResource:

```shell
[root@ebelarte chart]# oc create -f ice-kmod-sr.yaml 
specialresource.sro.openshift.io/ice-kmod created
[root@ebelarte chart]# 
```

And finally make sure that driver container is running:

```shell
[root@ebelarte chart]# oc get po 
NAME                                               READY   STATUS      RESTARTS   AGE
ice-kmod-driver-build-1-build                      0/1     Completed   0          53m
ice-kmod-driver-container-e383247e62b56585-pphbs   1/1     Running     0          45m
```

> ## EXERCISE 4

Reviewer: ```suggestion
## Deploy a simple driver-container as well as a device plugin
```

helm repo index cm --url=cm://dp-simple-kmod/dp-simple-kmod-chart 
oc create cm dp-simple-kmod-chart --from-file=cm/index.yaml --from-file=cm/dp-simple-kmod-0.0.1.tgz -n dp-simple-kmod
```

And deploy the new **SpecialResource** object:

```
oc create -f dp-simple-kmod-sr.yaml
```

### Verify both the daemonset and device plug-in are deployed and running in cluster

After a while we can check that the driver was built in-cluster by the **BuildConfig**, and the device-plugin and driver-container pods are running:

```
[root@ebelarte chart]# oc get po
NAME                                                     READY   STATUS      RESTARTS   AGE
dp-simple-kmod-deviceplugin-e383247e62b56585-qn2fb       1/1     Running     0          48m
dp-simple-kmod-driver-build-e383247e62b56585-1-build     0/1     Completed   0          50m
dp-simple-kmod-driver-container-e383247e62b56585-wrktm   1/1     Running     0          48m
```

And in our very specific example we could check if our “dummy” devices are present in the driver-container:

```
[root@ebelarte chart]# oc exec -it dp-simple-kmod-driver-container-e383247e62b56585-wrktm -- printenv | grep DUMMY
DUMMY_DEVICES=dev_1
```

> ## EXERCISE 5

Reviewer: ```suggestion
## Troubleshoot failures to load a driver-container
```
